Domain Records (#165)
dappforce/subsocial-parachain: pallets/domains/README.md

The main storages of domains pallet are: 
- `RegisteredDomains: Map<DomainName,DomainMeta>`\
  Metadata associated with a domain.
- `DomainsByOwner: Map<AccountId, DomainName[]>`\
  Domains owned by an account.
- `DomainRecords: Map<DomainName, Map<RecordKey, RecordValueWithDeposit>>`\
  All records associated with a given domain name.

## Domain Records

In addition to resolving a domain owner, you can also resolve the various records set up
by the user.

The domains record system allows any dotsama domain to have a key-value pairs of metadata attached to it.
There is no restrictions on what value should be stored in what key, however we have a list of defined key schema to be 
interpreted by clients.

### Record Key standard

A standard record key is split by namespaces by a . used as a separator.

#### Crypto payment records

> A key of records regarding payment in crypto follows one of the following formats

Author: @Yung-Beef thoughts on how we can make it better?

* `crypto.<TICKER>.address`
  * `crypto.ETH.address` ⇒ `0xD1E5b0FF1287aA9f9A268759062E4Ab08b9Dacbe`
  * `crypto.BTC.address` ⇒ `bc1qkd4um2nn2uyzmsch5y86wsa2pfh8xl445lg9nv`
* `crypto.<TICKER>.version.<VERSION>.address`
  * `crypto.USDT.version.ERC20.address` ⇒ `0x8aaD44321A86b170879d7A244c1e8d360c99DdA8`
  * `crypto.USDT.version.TRON.address` ⇒ `THG9jVSMfKEbg4vYTYWjmLRyga3CKZdDsk`

#### Browser resolution

* `browser.redirect_url` ⇒ `http://example.com/home.html`

#### Social records

* `social.picture.value` ⇒ `ipfs://QmQqzMTavQgT4f4T5v6PWBp7XNKtoPmC9jvn12WPT3gkSE`
* `whois.email.value` ⇒ `tarekkma@gmail.com`
* `whois.for_sale.value` ⇒ `false`
* `social.twitter.username` ⇒ `TarekkMA1`

> * `social.subsocial.account` ⇒ `3tiKakuy6RaHWThsHts23De8XwpaNdvw8PwwXcJqeVSN6w8w`

Author: You can manage your domain from a separate account from your main account.

* `social.subsocial.space` ⇒ `15`

> * `social.subsocial.post` ⇒ `1`

Author: The domain owner can be different than this record value. and for consistsity we make all social records follow the same style

### Main types

> * `DomainName`: an array of characters representing a domain name (maximum length `Config::MaxDomainLength`)
> * `RecordKey`: an array of characters representing record key (maximum length `Config::MaxRecordKeySize`)
> * `RecordValue`: an array of characters representing record value (maximum length `Config::MaxRecordValueSize`)

Reviewer: These descriptions sound too noisy. The names of the types are self-descriptive.
```suggestion
* `DomainName` – A byte array of maximum length of `Config::MaxDomainLength` bytes.
* `RecordKey` – A byte array of maximum length of `Config::MaxRecordKeySize` bytes.
* `RecordValue` – A byte array of maximum length of `Config::MaxRecordValueSize` bytes.
```

```rust

> pub struct DomainMeta<T: Config> {

Author: I don't know what you mean

  Metadata associated with domain.
- `DomainsByOwner: Map<AccountId, DomainName[]>`\
  Domains owned by account.
- `DomainRecords: Map<DomainName, Map<RecordKey, RecordValueWithDeposit>>`\
  All records associated with a given domain name.

## Domain Records

In addition to resolving a domain owner, you can also resolve the various records set up
by the user.

Domains records system allows any dotsama domains to have key-value pairs of metadata attached to it.
There is no restrictions on what value should be stored in what key, However we have a list of defined key schema to be 
interpreted by clients.

### Record Key standard

> Standard record key is split by namespaces by a . used as a separator.

Reviewer 2: ```suggestion
A standard record key is split by namespaces by a . used as a separator.
```

Reviewer 2: I don't really understand what this means, can you give an example and I can try to rewrite this line more clearly?

Author: In the record system, anyone can just put jibberish as a key. example `key:"MyAwsomeName" => value:"Beef"`.

This flexibility is great but we want some sort of standardization so different clients from different teams can interpret and use these records in a meaningful way. That's why I suggest that we split the key into namespaces that are split by dot (`social.name => Tarek`)

#### Crypto payment records

> key of records regarding payment in crypto follows one of the following formats

Reviewer 2: ```suggestion
A key of records regarding payment in crypto follows one of the following formats
```

* `crypto.<TICKER>.address`
  * `crypto.ETH.address` ⇒ `0xD1E5b0FF1287aA9f9A268759062E4Ab08b9Dacbe`
  * `crypto.BTC.address` ⇒ `bc1qkd4um2nn2uyzmsch5y86wsa2pfh8xl445lg9nv`
* `crypto.<TICKER>.version.<VERSION>.address`
  * `crypto.USDT.version.ERC20.address` ⇒ `0x8aaD44321A86b170879d7A244c1e8d360c99DdA8`
  * `crypto.USDT.version.TRON.address` ⇒ `THG9jVSMfKEbg4vYTYWjmLRyga3CKZdDsk`

#### Browser resolution

* `browser.redirect_url` ⇒ `http://example.com/home.html`

#### Social records

* `social.picture.value` ⇒ `ipfs://QmQqzMTavQgT4f4T5v6PWBp7XNKtoPmC9jvn12WPT3gkSE`
* `whois.email.value` ⇒ `tarekkma@gmail.com`
* `whois.for_sale.value` ⇒ `false`
* `social.twitter.username` ⇒ `TarekkMA1`
* `social.subsocial.account` ⇒ `3tiKakuy6RaHWThsHts23De8XwpaNdvw8PwwXcJqeVSN6w8w`
* `social.subsocial.space` ⇒ `15`
* `social.subsocial.post` ⇒ `1`

### Main types

> * `DomainName`: array of characters representing a domain name (maximum length `Config::MaxDomainLength`)

Reviewer 2: ```suggestion
* `DomainName`: an array of characters representing a domain name (maximum length `Config::MaxDomainLength`)
```

Add "an" to the following 2 lines

    /// The domain owner.
    owner: T::AccountId,

>     /// The amount was held as a deposit for storing this structure.

Reviewer 2: ```suggestion
    /// The amount that was held as a deposit for storing this structure.
```
